### DISCUSSION TITLE
ADPiecewiseLinearInterpolationMaterial Question

### DISCUSSION THREAD

**Discussion by jsevic** [idaholab/moose#19631] [Q&A Modules: General] [answered]
I was using the ElectricalConductivity material property to compute and return electrical conductivity for both of these kernels below. If my understanding is correct, JouleHeatingSource looks for this material property and HeatConduction in this case is used as the diffusion operator on electrical_conductivity, which in contrast to JouleHeatingSource, must be specified. This sim works fine.

```
  [./HeatSrc]
    type = JouleHeatingSource
    variable = T
    elec = V
  [../]
  [./electric]
    type = HeatConduction
    variable = V
    diffusion_coefficient = electrical_conductivity
  [../]

  [./ElectricalConductivity]
    type = ElectricalConductivity
    temperature = T
    block = 0
    reference_resistivity = 0.001
    reference_temperature = 293
    temperature_coefficient = 1e-6
  [../]
```

Now I would like instead of assuming a functional expression for electrical_conductivity(T) is to specify (x,y) pairs of T and sigma such that MOOSE will linearly interpolate between these (x,y) pairs. I did this successfully earlier with a custom ExampleMaterial

```
  [./ElectricalMaterialProperties_TaO]
    type = ExampleMaterial
    block = 'TaO'
    independent_vals = '100 200 300 400 500'
    dependent_vals = '4e2 1 0 -1e2 -1e3'
    #dependent_vals = '10e2 10e2 10e2 10e2 10e2'
  [../]
```

I tried the ADPiecewiseLinearInterpolationMaterial to interpolate electrical conductivity that is used in the 

```
  [./ElectricalConductivity_TaO]
    type = ADPiecewiseLinearInterpolationMaterial
    xy_data = '100 4e2
               200 1
               300 -1e2
               400 -1e3
               1500 -2e3'
    property = electrical_conductivity
    variable = T
    block = 'TaO'
  [../]
```

and I get the following error...

Attempting to declare AD material property electrical_conductivity, but it is already retrieved/declared as a regular material property.

How do I tell JouleHeatingSource what electrical conductivity to use?

Thx!

**Comment by friedmud** [comment]
Do you still have the `ElectricalConductivity` material there too?  This is saying that you have another _Material_ that has already declared a non-AD version of that property.

**Reply by jsevic** [reply]
Doesn't look like it, I have several materials of type ElectricalConductivity except for this material I used ADPiecewiseLinearInterpolationMaterial. I do use GenericConstantMaterial and HeatConductionMaterial on this same material. Do these specify electrical_conductivity as well?

**Comment by jsevic** [comment]
Just to test, I replaced that material with a generic material and it runs.

  [./ElectricalConductivity_TaO]
    type = ElectricalConductivity
    temperature = T
    block = 'TaO'
    reference_resistivity = 1e-4
    reference_temperature = 293
    temperature_coefficient = 0
  [../]

**Comment by GiudGiud** [comment, marked as answer]
Hello

You can change the name used for electrical_conductivity in JouleHeatingSource by adding the parameter
`electrical_conductivity = new_name`

What's happening right now is that JouleHeatingSource is the non AD version of that kernel. So it retrieves the non-AD version of electrical_conductivity. ADPiecewiseInterpolationMaterial defines an AD property. If you want to use AD, use ADJouleHeatingSource (and ADElectricalConductivity) 

Guillaume

**Comment by jsevic** [comment]
Ok thanks let me try. Could I make this a variable that is used elsewhere?

**Reply by GiudGiud** [reply]
Hello

Not sure what you mean.
If you mean rename it everywhere, then you can set `electrical_conductivity = new_name` in the [GlobalParams] block to pass this parameter to every object that will accept it. If you get unexpected failures from this, you can pass `--show-input` on the command line to see what the input file looks like with all the additional parameters added. 

If you mean use an auxiliary variable instead of a material property, I m afraid a lot of the objects you are using in your input expect a material property, and conductivities are usually classified as material properties, so it's better not to do that.

**Comment by jsevic** [comment]
Ok and what about if I want to pass electrical_conductivity to the heat diffusion operator, for example, would that be ok?

**Reply by GiudGiud** [reply]
This one?
https://mooseframework.inl.gov/source/kernels/HeatConduction.html
If so then set `diffusion_coefficient = electrical_conductivity_name` (the new name if you are modifying it)

**Comment by jsevic** [comment]
Yeah that’s it - cool thx let me try.

**Comment by jsevic** [comment]
Another question - not sure if I should start another thread. I have an electrical conductivity model included in my current app, using the semiconductor model fitted to my data.

Now earlier in my phase field app, I used the CH split parsed kernel and the state variables were w and c. For initial testing, I used c as a proxy for conductivity, and coupled it to a custom diffusion operator kernel that allowed me to pass c. In the current MOOSE release in the heat conduction app there is supplied a diffusion operator kernel that does what my older custom kernel did, so I have been using this new one. I also use it for charge conservation and pass to it electrical_conducrivity and this works as expected, I have replicated my older electrothermal code using the current release of MOOSE.

Now my question is, the app uses electrical_conductivity in two places - the charge conservation kernel and the Joule heating kernel. Since I want to scale electrical_conductivity by c, the concentration from the CH kernel, can I simply replace electrical_conductivity with the product c*electrical_conductivity in the electrical conductivity declaration of each of these two diffusion operator kernels, If not, what should I try?

Thank you!

**Reply by GiudGiud** [reply]
So you want to pass 'c*electrical_conductivty' in the arguments?
This wont get parsed automatically. 
I would create a `ParsedMaterial` with the new definition for the conductivity, then give the name of that new material property in lieu of the conductivities. https://mooseframework.inl.gov/docs/PRs/19934/site/source/materials/ParsedMaterial.html

**Comment by jsevic** [comment]
Ok cool let me try thx!

**Comment by jsevic** [comment]
I have a couple of more questions related to this thread, let me know if you want me to start a new thread.

Q1.
My current state variables are T, V, c, and w for our phase-field app. I would like to plot current density. Right now I have defined an auxiliary variable rho_e and use it with an auxiliary kernel to produce an output variable for rho_e. If I were to do this approach for computing current density, I would need grad V, is there a way to do this without creating a new (compiled?) kernel?

  [./ElectricalConductivity]
    # (S/m)
    # Create our linear electrical conductivity model here to plot.
    type = ParsedAux
    variable = rho_e
    args = 'T c'
    function = '(800 + 0.8*T^4)*c*(1-c)'
  [../]


Q2.
For my current IC c(x,y,0) for concentration, I would like it to be uniform between 0 and 0.3 and 0.7 and 1.0 but 0 in between, is there a way to do this? I saw something called distributions?


Thank you!

**Reply by GiudGiud** [reply]
Hello

Q1
So to output the gradient of a variable you can use this:
https://mooseframework.inl.gov/moose/source/auxkernels/VariableGradientComponent.html
and place the result in each component. 

Q2
Yes certainly. 
I think the easiest way will be to use a ParsedFunction and then use a FunctionIC

Guillaume
